edgelet/edgelet-http-hosting/src/client/hosting.rs

```diff
@@ -0,0 +1,219 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+use std::sync::Arc;
+
+use failure::{Fail, ResultExt};
+use futures::prelude::*;
+use hosting::apis::client::APIClient;
+use hosting::apis::configuration::Configuration;
+use hosting::apis::HostingApi;
+use hosting::models::*;
+use hyper::Client;
+use url::Url;
+
+use edgelet_core::UrlExt;
+use edgelet_http::UrlConnector;
+
+use crate::error::{Error, ErrorKind};
+
+pub trait HostingInterface {
+    type Error: Fail;
+
+    type DeviceConnectionInformationFuture: Future<Item = DeviceConnectionInfo, Error = Self::Error>
+        + Send;
+
+    fn get_device_connection_information(&self) -> Self::DeviceConnectionInformationFuture;
+}
+
+pub trait GetApi {
+    fn get_api(&self) -> &dyn HostingApi;
+}
+
+pub struct HostingClient {
+    client: Arc<dyn GetApi>,
+}
+
+impl GetApi for APIClient {
```

Comment:
Why do we need this?

Reply:
For unit tests below. I needed a way to mock the underlying HostingApi so that I can mock errors in the tests and assert that they're being handled correctly.

---
In reply to: [282090767](https://github.com/Azure/iotedge/pull/1144#discussion_r282090767) [](ancestors = 282090767)